File: server/routes/hole_detect.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from server.bundles.storage import get_bundle
from server.security import require_api_key
from server.services.hole_detect import SuggestedHole, suggest_hole_for_location

router = APIRouter(dependencies=[Depends(require_api_key)])


class HoleDetectQuery(BaseModel):
    courseId: str
    lat: float
    lon: float
    lastHole: int | None = None


class HoleDetectResponse(BaseModel):
    hole: int
    distance_m: float
    confidence: float
    reason: str


@router.post("/api/hole/detect", response_model=HoleDetectResponse)
async def detect_hole(body: HoleDetectQuery) -> HoleDetectResponse:
    bundle = get_bundle(body.courseId)
    if bundle is None:
        raise HTTPException(status_code=404, detail="Unknown courseId")

    suggestion: SuggestedHole | None = suggest_hole_for_location(
        bundle=bundle,
        lat=body.lat,
        lon=body.lon,
        last_hole=body.lastHole,
    )
    if suggestion is None:
        raise HTTPException(status_code=404, detail="No suitable hole found")

    return HoleDetectResponse(
        hole=suggestion.hole,
        distance_m=suggestion.distance_m,
        confidence=suggestion.confidence,
        reason=suggestion.reason,
    )

File: web/src/api/holeDetect.ts
import axios from "axios";

import { API, withAuth } from "@/api";

export type HoleDetectRequest = {
  courseId: string;
  lat: number;
  lon: number;
  lastHole?: number | null;
};

export type HoleDetectResponse = {
  hole: number;
  distance_m: number;
  confidence: number;
  reason: string;
};

export async function detectHole(
  body: HoleDetectRequest,
): Promise<HoleDetectResponse> {
  const resp = await axios.post<HoleDetectResponse>(`${API}/api/hole/detect`, body, {
    headers: withAuth({ "Content-Type": "application/json" }),
  });

  return resp.data;
}
